### ISSUE TITLE
Multiple Paypal buttons throws error

### ISSUE THREAD

**Issue by prindacerk** [Enngage/ngx-paypal#46]
I have 3 separate components that display the Paypal pay button in the same web page. I noticed when I render them together, it throws an exception.
`
"Error: zoid desroyed all
    at no (https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:100969)
    at Ra (https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:186269)
    at https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:189023
    at Array.forEach (<anonymous>)
    at Object.value (https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:188996)
    at https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:188375
    at Module.<anonymous> (https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:189062)
    at t (https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:157)
    at https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:941
    at https://www.paypal.com/sdk/js?client-id=xxxxx&currency=AUD&commit=true:1:950"
`
FYI, I do conditionally show them using *ngif  if the user has logged in. So while the initialization happens ngOnInit, the display of the button on screen happens conditionally.

**Comment by prindacerk** [comment]
From what I was able to find out from the Paypal dev, this is caused by the script loading multiple times. He suggested that I don't do that and only load it once. Then this problem will be resolved.

Is there anyway you can design your component to load the script only if it's not loaded already?

**Comment by Enngage** [comment]
This is kind of problematic because the new JavaScript SDK expects certain parameters to be configured via query params in URL and so if you have different configuration, having multiple PayPal buttons will never work with the current implementation of PayPal's SDK. I could possibly let you register the script once and if you share the base configuration, it might work for some usecases.

**Comment by Enngage** [comment]
I added the ability to the above, can you try that as per docs here - https://github.com/Enngage/ngx-paypal#multiple-ngx-paypal-components-on-the-same-page 

Let me know how that goes!

**Comment by prindacerk** [comment]
@Enngage Thanks for that approach. However, it sort of makes the components dependent of one another in the page. If the web page is broken by multiple components and two separate components have the Paypal window within them, then this will be tedious to handle.

A simple use case. Let's say that I have a pricing comparison page similar to this.
https://coderthemes.com/abstack/default/extras-pricing.html
Now each pricing package will be a separate component. And the button listed in that package is the ngx-paypal button. I shouldn't be tying one package's button to the other component's button. It creates a dependency there. And more packages I add, the more I would have to create dependencies for.

How about modifying your component so you assume that the script will be loaded by the user in their html page. And just implement the functionality with the assumption that the script exist? Make the script inclusion as one of the dependencies and let the user implement that while you can do the functionality of setting up and triggering the actions?

**Comment by Enngage** [comment]
Yeah, I guess thats a fair point. How about now?

**Comment by prindacerk** [comment]
Hi mate,
Apologies for the delay in replying.

I ended up creating a mockup of my solution (to provide to Paypal for another issue). Maybe it will help you understand and figure out how it should be implemented?.
I didn't use your component but used the Paypal script directly instead. I did that so it gets loaded only once. I would prefer to let the loading of the script to be handled by the user and your component help simplify the Paypal component building part.
https://github.com/prindacerk/paypal.test

Now if your component on loading were to check if the Paypal property is defined or not and if not, then define itself or else it will use existing global variable, then the dependency on the script will be resolved. User can choose to either load the script themselves or load it via your component. Your component can focus more on creating the objects to handle the callbacks and response which in my example is built on javascript.

**Comment by Enngage** [comment]
Yeah, thats what you can do with the new implementation as described here - https://github.com/Enngage/ngx-paypal#multiple-ngx-paypal-components-on-the-same-page 

You can instruct all components to not register script, and call `customInit` with the payPal api once that is available. In the example I'm using the `PayPalScriptService` which would be recommended way, but you can just dump the file onto your index.html and get paypal from `window` and pass that through `customInit`

**Comment by prindacerk** [comment]
@Enngage I did see that. But aren't you making the components (paypalComponent1, paypalComponent2) dependent of YourComponent? YourComponent would have to be aware of all the paypal components in this scenario right?

Also, by doing this way, you are assuming that both components are driven by the same configuration. In my example, you would see that all 3 components have different configs because the price and details are different. So it's multiple paypal buttons in same page and they have different configs. Whether it's same config or different config can be a decision to the end user.

My demo works fine on it's own. It renders the buttons in the same page and they are different components shown on the same page. The page doesn't need to be aware of the other buttons as they can work independently of one another. Have a look and let me know what you think.

**Comment by Enngage** [comment]
You can use different config for each of the components - I only used one for simplicity. I'm not sure what do you mean by making paypalComponent1 (and 2) dependent on YourComponent. This is how Angular works :)

**Comment by prindacerk** [comment]
What I mean is that you are injecting @ViewChild into YourComponent and then setting the Paypal API to those child components correct? So YourComponent is passing the API to the child components. That's what I meant by dependency.
If you look at my example, https://github.com/prindacerk/paypal.test/blob/master/src/app/packages/list.component.html
shows the Paypal components. But the related component https://github.com/prindacerk/paypal.test/blob/master/src/app/packages/list.component.ts does not have to configure anything. It just plays as a placeholder. The PaypalConfig is all within the component that renders the component. Ex: 
https://github.com/prindacerk/paypal.test/blob/master/src/app/packages/_basic.component.ts.

Now let me ask this. Is it possible to call this.payPalScriptService.registerPayPalScript multiple times within a page from different components? Would it interfere with one another? If it can be called multiple times independently, then we are good. Then it can be implemented within the component that shows the Paypal button. As it is in your example, it is used in parent component which passes it to the child component where the question of dependency comes in.

**Comment by Enngage** [comment]
The PayPal script cannot be used multiple times on the same page - that is the way JavaScript SDK works so you couldn't really  register scripts using `registerPayPalScript`  multiple times, but you really don't need to do that. 

You can have shared service in your components with observable that received the payPal API in its payload and pass this to your components. You will then register the script from the 'most top (parent)' component that you can. All your components will subscribe to this service and receive payPal API at the same time.

**Comment by prindacerk** [comment]
> You will then register the script from the 'most top (parent)' component that you can.
Would this require that to be called in the AppComponent? Since that's the top root component right?

**Comment by Enngage** [comment]
That is one of the possibilities. You could implement some sort of semaphor so that each of your components try to register the script and whichever is first will broadcast it and stop all others. 

Or you can just include the paypal javascript in your index.html and get it from window object.

**Comment by prindacerk** [comment]
@Enngage I managed to implement your component in my example.
https://github.com/prindacerk/paypal.test/tree/ngx-paypal
Please have a look (especially the app.component and  _basic.component files. See if I have implemented as you suggest I do.

One minor issue I had is that eventhough I use the service and create the Api which I pass to the customInit, I still need to give the ClientId and Currency in the PaypalConfig. It won't work without those fields. Is there anyway you can make them optional?

Also, I noticed that your PaypalConfig's OnClick  doesn't allow parameters (data, actions) to be passed. They have implemented validation logic at this stage so allowing parameters in this method would really be useful.

**Comment by Enngage** [comment]
I had a quick look and what you have is not exactly what I had in meant. If you look above, the idea is that you use my PayPalScriptService to register PayPal's javascript SDK (or you dump it on your index.html) and once the script loads, you will be able to access it through global variable (window['paypal']). Once you have this, you can then call `customInit` method on any  of your ngx-paypal components and pass the payPal API to it.

**Comment by prindacerk** [comment]
Sorry. I don't understand. Isn't that what I did?
I registered the SDK using your **PayPalScriptService** in the AppComponent which is the most top component. Then I stored the generated PaypalApi in my own PackageService so it is available anywhere. Then in the BasicPackageComponent, I retrieve the PaypalApi from my service and pass it to the **customInit**.
